### PR TITLE
Measurement scripts tested and fixed issues

### PR DESCRIPTION
**Pull request by raghavagrawal**
scripts are configured to run from BITS-Darshini directory. A folder named darshini-logs must be created in the directory. In darshini-test.sh <Path to BITS-Darshini> must be replaced with the relevant path. ntopng is still not working.

### CODE REVIEW

File: scripts/measurements/tools_run.sh [hunk 1]
```diff
@@ -8,37 +8,34 @@
...
 
+truncate -s 0 performance.log
 echo -e "bro performance\n" >> performance.log
-for pcapFile in `ls Test_pcap_files/sample_capture_test*.pcap`
+for pcapFile in `sudo ls data/packet/*.pcap`
 do
 
     #perform a trial run to get all the data into RAM
-    /usr/bin/time bro -r $pcapFile > /dev/null 2>&1
+    sudo /usr/bin/time bro -r $pcapFile > /dev/null 2>&1
```
Comment (prasadtalasila) [inline scripts/measurements/tools_run.sh]: It would be better to run this command without a sudo? Would we be able to do it?

File: scripts/measurements/tools_run.sh [hunk 2]
```diff
@@ -8,37 +8,34 @@
 PATH=$PATH:/opt/bro/bin
 export PATH
 
+truncate -s 0 performance.log
 echo -e "bro performance\n" >> performance.log
-for pcapFile in `ls Test_pcap_files/sample_capture_test*.pcap`
+for pcapFile in `sudo ls data/packet/*.pcap`
```
Comment (prasadtalasila) [inline scripts/measurements/tools_run.sh]: It would be better to do away sudo here.

File: scripts/measurements/tools_run.sh (hunk 1 above)
Comment (prasadtalasila) [reply scripts/measurements/tools_run.sh]: same comments holds for sudo in all the following lines in this script.

File: scripts/measurements/tools_run.sh [hunk 3]
```diff
@@ -8,37 +8,34 @@
 PATH=$PATH:/opt/bro/bin
 export PATH
 
+truncate -s 0 performance.log
```
Comment (prasadtalasila) [inline scripts/measurements/tools_run.sh]: It would be better to put all the log files in BITS-Darshini/data/log/ directory. It is also better to create that directory if it does not exist. Assuming the existence of that directory is going to lead to lot of unintended errors.

File: scripts/measurements/darshini-test.sh [hunk 4]
```diff
@@ -10,10 +10,10 @@
 
 truncate -s 0 darshini-test.log
 
-echo "performance of Darshini on four processors"  >> darshini-test.log
+
 echo "--------------------------------------" >> darshini-test.log
 
-for pcapFile in `ls Test_pcap_files/sample_capture_test*.pcap`
+for pcapFile in `sudo ls <Path to BITS-Darshini>/data/packet/*.pcap`
```
Comment (prasadtalasila) [inline scripts/measurements/darshini-test.sh]: sudo not needed here. We can give a relative path name here (same way the path way provided in tools_run.sh script).

File: scripts/measurements/darshini-test.sh [hunk 5]
```diff
@@ -28,7 +28,7 @@ do
     echo -e "\n\n"  >> darshini-test.log
     #run one experiment in Darshini
     curl -H "Content-Type: application/json" -d '{"email": "abc", "password": "abc"}' http://localhost:8080/protocolanalyzer/signin
-    curl -X GET -H "Content-Type: application/json" http://localhost:8080/protocolanalyzer/test?pcapPath=\'$pcapFile\'
+    curl -X GET -H "Content-Type: application/json" http://localhost:8080/protocolanalyzer/test?pcapPath=$pcapFile
```
Comment (prasadtalasila) [inline scripts/measurements/darshini-test.sh]: the single quotes protect files names with spaces and special characters in them. It might be better to keep them.

Comment (prasadtalasila) [reply scripts/measurements/darshini-test.sh]: again, the log files must be put in BITS-Darshini/data/log directory.

**Review by prasadtalasila — CHANGES_REQUESTED** [review]
@raghavagrawal please checked the suggested changes and make them if appropriate.

Do create the BITS-Darshini/data/log directory and keep it empty at present. You may need to add an empty .gitkeep file in that directory in order to push the directory to the git repository

File: scripts/measurements/tools_run.sh (hunk 1 above)
Comment (raghavagrawal) [reply scripts/measurements/tools_run.sh]: Running bro without sudo gives an error.

Comment (raghavagrawal) [reply scripts/measurements/tools_run.sh]: @prasadtalasila Also, sir, this script must be run as a non-root user because of Wireshark. Due to running it as a non-root user, there is a permission denied error when trying to write into the performance.log file. This can be removed using chmod command. So, can I include this command in this script?

Comment (prasadtalasila) [reply scripts/measurements/tools_run.sh]: bro should run without sudo. Can you please check it again to see if the file permissions is the issue here?

**Comment by prasadtalasila** [comment]
@raghavagrawal you can make the other changes apart from bro. Rest of the script should be able to run without sudo. Suitable changes can be made to them. 

As mentioned in the previous comment today, even bro should run without sudo permission; please check the file permissions issue if any.

**Comment by raghavagrawal** [comment]
@prasadtalasila It was a file permission issue. Bro is running without root user. The problem is with log file permission. And also when concurrency scripts are run as non-root user need permission to access  
/sys/devices/system/cpu/ folder.

**Comment by prasadtalasila** [comment]
@raghavagrawal I agree that /sys/devices/system/cpu/ folder needs sudo permission. We can use a modified command
```bash
> sudo bash -c 'echo 1 > /sys/devices/system/cpu/cpu1/online'
```
to solve the problem.

File: scripts/measurements/darshini-test.sh (hunk 4 above)
Comment (raghavagrawal) [reply scripts/measurements/darshini-test.sh]: @prasadtalasila Relative path name here is not working as we have to give the path name to the analyzer in BITS-Darshini which requires the whole path name. I think that we can maybe add pwd command in the script and store it result in a variable and use that in ls command.

File: scripts/measurements/darshini-test.sh (hunk 5 above)
Comment (raghavagrawal) [reply scripts/measurements/darshini-test.sh]: @prasadtalasila the script with single quotes is not working properly.

**Comment by raghavagrawal** [comment]
@prasadtalasila changes are done. There are some changes in darshini-test.sh which could not be completed.

File: scripts/measurements/darshini-test.sh [hunk 6]
```diff
@@ -3,34 +3,33 @@
...
+    echo -e "\n\n\nExperiment"  >> data/log/darshini-test.log
+    echo -e "---------------" >> data/log/darshini-test.log
+    echo -e "file name = $pcapFile, number of cpus=$cpus" >> data/log/darshini-test.log
+    echo -e "\n\n"  >> data/log/darshini-test.log
     #run one experiment in Darshini
     curl -H "Content-Type: application/json" -d '{"email": "abc", "password": "abc"}' http://localhost:8080/protocolanalyzer/signin
-    curl -X GET -H "Content-Type: application/json" http://localhost:8080/protocolanalyzer/test?pcapPath=\'$pcapFile\'
+    curl -X GET -H "Content-Type: application/json" http://localhost:8080/protocolanalyzer/test?pcapPath=$pcapFile
     sleep 10
-    cat darshini-logs/darshini >> darshini-test.log
```
Comment (prasadtalasila) [inline scripts/measurements/darshini-test.sh:33]: @raghavagrawal why is this line removed? The line helps collect the the analysis log from BITS Darshini webapp into darshini-test.log

File: scripts/measurements/darshini-test.sh [hunk 7]
```diff
@@ -3,34 +3,33 @@
...
-    echo -e "file name = $pcapFile, number of cpus=$cpus" >> darshini-test.log
-    echo -e "\n\n"  >> darshini-test.log
+    echo -e "\n\n\nExperiment"  >> data/log/darshini-test.log
+    echo -e "---------------" >> data/log/darshini-test.log
+    echo -e "file name = $pcapFile, number of cpus=$cpus" >> data/log/darshini-test.log
+    echo -e "\n\n"  >> data/log/darshini-test.log
     #run one experiment in Darshini
     curl -H "Content-Type: application/json" -d '{"email": "abc", "password": "abc"}' http://localhost:8080/protocolanalyzer/signin
-    curl -X GET -H "Content-Type: application/json" http://localhost:8080/protocolanalyzer/test?pcapPath=\'$pcapFile\'
+    curl -X GET -H "Content-Type: application/json" http://localhost:8080/protocolanalyzer/test?pcapPath=$pcapFile
```
Comment (prasadtalasila) [inline scripts/measurements/darshini-test.sh:30]: why were the single quotes removed?

File: scripts/measurements/tools_run.sh [hunk 8]
```diff
@@ -18,27 +19,23 @@ do
...
     #/usr/bin/time  --format "$pcapFile,%S,%M" bro -r $pcapFile
     #actual run with output of elapsed time and max memory
-    /usr/bin/time  --format "$pcapFile,%e,%M" bro -r $pcapFile
-done >> performance.log 2>&1
-
-
-echo -e "\n--------------------------\n" >> performance.log
-echo -e "ntopng performance\n" >> performance.log
-for pcapFile in `ls Test_pcap_files/sample_capture_test*.pcap`
+    sudo /usr/bin/time  --format "$pcapFile,%e,%M" bro -r $pcapFile
```
Comment (prasadtalasila) [inline scripts/measurements/tools_run.sh:22]: why do we need sudo here?

**Review by prasadtalasila — CHANGES_REQUESTED** [review]
Please look at the comments and suggest suitable reasons

File: scripts/measurements/darshini-test.sh (hunk 7 above)
Comment (adityaketkar) [reply scripts/measurements/darshini-test.sh:30]: We tested the script, it was working only without the quotes for us.

File: scripts/measurements/darshini-test.sh (hunk 6 above)
Comment (raghavagrawal) [reply scripts/measurements/darshini-test.sh:33]: I will again put in this line. This was some misunderstanding.

File: scripts/measurements/tools_run.sh (hunk 8 above)
Comment (raghavagrawal) [reply scripts/measurements/tools_run.sh:22]: @prasadtalasila without sudo there was a warning message 

>warning: can't stat directory .state/.tmp: Permission denied

**Comment by raghavagrawal** [comment]
@prasadtalasila Can you please review the comments.

**Comment by prasadtalasila** [comment]
@raghavagrawal @adityaketkar I have refactored the code in raghavagrawal:measurements branch and pushed to [measurements-refactor](https://github.com/prasadtalasila/BITS-Darshini/tree/measurements-refactor) branch. 
I am yet to verify the functionality completely by running the code completely. I will do that in a day. In the mean time, you can also verify the functionality of the refactored branch. Thanks.

**Comment by adityaketkar** [comment]
All tests work fine for me, except for the ntopng error that we discussed previously.
Here are the contents of concurrency.log containing test results for my laptop.

`performance of tools on single processor
--------------------------------------
bro performance

data/packet/DNS_Traffic000.pcap,0.21,41368
data/packet/packet_data.pcap,0.22,46124

--------------------------

ntopng performance

Command terminated by signal 6
data/packet/DNS_Traffic000.pcap,13.43,63936
Command terminated by signal 6
data/packet/packet_data.pcap,13.25,61584

--------------------------

tshark performance

data/packet/DNS_Traffic000.pcap,4.80,200292
data/packet/packet_data.pcap,0.21,202508




--------------------------------------
performance of tools on two processors
--------------------------------------
bro performance

data/packet/DNS_Traffic000.pcap,0.20,46204
data/packet/packet_data.pcap,0.17,43612

--------------------------

ntopng performance

Command terminated by signal 6
data/packet/DNS_Traffic000.pcap,4.42,63912
Command terminated by signal 6
data/packet/packet_data.pcap,4.42,64344

--------------------------

tshark performance

data/packet/DNS_Traffic000.pcap,0.22,202192
data/packet/packet_data.pcap,0.20,201660




--------------------------------------
performance of tools on three processors
--------------------------------------
bro performance

data/packet/DNS_Traffic000.pcap,0.20,46276
data/packet/packet_data.pcap,0.23,41504

--------------------------

ntopng performance

Command terminated by signal 6
data/packet/DNS_Traffic000.pcap,4.45,63648
Command terminated by signal 6
data/packet/packet_data.pcap,4.45,67976

--------------------------

tshark performance

data/packet/DNS_Traffic000.pcap,0.23,202076
data/packet/packet_data.pcap,0.20,200856




--------------------------------------
performance of tools on four processors
--------------------------------------
bro performance

data/packet/DNS_Traffic000.pcap,0.19,44860
data/packet/packet_data.pcap,0.17,46392

--------------------------

ntopng performance

Command terminated by signal 6
data/packet/DNS_Traffic000.pcap,4.47,64264
Command terminated by signal 6
data/packet/packet_data.pcap,4.45,63824

--------------------------

tshark performance

data/packet/DNS_Traffic000.pcap,0.22,201100
data/packet/packet_data.pcap,0.22,202004`

**Comment by prasadtalasila** [comment]
I ran all the tests on my laptop. They seem to run fine. I am closing this PR and am going to merge measurements-refactor branch on to dev. Good work @raghavagrawal and @adityaketkar
